src/ImageSharp/Formats/Jpg/JpegEncoderCore.cs

```diff
@@ -534,122 +620,84 @@ private int WriteBlock(ref Block block, QuantIndex index, int prevDC)
+                        huffman[len++] = count[c];
+                    }
+
+                    for (int v = 0; v < spec.Values.Length; v++)
+                    {
+                        huffman[len++] = values[v];
                     }
                 }
+
+                this.outputStream.Write(this.huffmanBuffer, 0, len);
             }
         }
 
         /// <summary>
-        /// Writes the application header containing the JFIF identifier plus extra data.
+        ///     Writes the Define Quantization Marker and tables.
         /// </summary>
-        /// <param name="horizontalResolution">The resolution of the image in the x- direction.</param>
-        /// <param name="verticalResolution">The resolution of the image in the y- direction.</param>
-        private void WriteApplicationHeader(short horizontalResolution, short verticalResolution)
+        private void WriteDefineQuantizationTables()
         {
-            // Write the start of image marker. Markers are always prefixed with with 0xff.
-            this.buffer[0] = JpegConstants.Markers.XFF;
-            this.buffer[1] = JpegConstants.Markers.SOI;
-
-            // Write the JFIF headers
-            this.buffer[2] = JpegConstants.Markers.XFF;
-            this.buffer[3] = JpegConstants.Markers.APP0; // Application Marker
-            this.buffer[4] = 0x00;
-            this.buffer[5] = 0x10;
-            this.buffer[6] = 0x4a; // J
-            this.buffer[7] = 0x46; // F
-            this.buffer[8] = 0x49; // I
-            this.buffer[9] = 0x46; // F
-            this.buffer[10] = 0x00; // = "JFIF",'\0'
-            this.buffer[11] = 0x01; // versionhi
-            this.buffer[12] = 0x01; // versionlo
-            this.buffer[13] = 0x01; // xyunits as dpi
-
-            // No thumbnail
-            this.buffer[14] = 0x00; // Thumbnail width
-            this.buffer[15] = 0x00; // Thumbnail height
-
-            this.outputStream.Write(this.buffer, 0, 16);
+            // Marker + quantization table lengths
+            int markerlen = 2 + (QuantizationTableCount * (1 + Block8x8F.ScalarCount));
+            this.WriteMarkerHeader(JpegConstants.Markers.DQT, markerlen);
 
-            // Resolution. Big Endian
-            this.buffer[0] = (byte)(horizontalResolution >> 8);
-            this.buffer[1] = (byte)horizontalResolution;
-            this.buffer[2] = (byte)(verticalResolution >> 8);
-            this.buffer[3] = (byte)verticalResolution;
+            // Loop through and collect the tables as one array.
+            // This allows us to reduce the number of writes to the stream.
+            int dqtCount = (QuantizationTableCount * Block8x8F.ScalarCount) + QuantizationTableCount;
+            byte[] dqt = ArrayPool<byte>.Shared.Rent(dqtCount);
+            int offset = 0;
 
-            this.outputStream.Write(this.buffer, 0, 4);
-        }
+            JpegUtils.WriteDataToDqt(dqt, ref offset, QuantIndex.Luminance, ref this.luminanceQuantTable);
```

Comment:
That's my rule as well. If anyone else has a need for it then it can move into a common place.